openedx/core/djangoapps/programs/utils.py

```diff
@@ -226,12 +264,38 @@ def completed_course_runs(self):
         Returns:
             list of dicts, each representing a course run certificate
         """
+        return self.course_runs_with_state['completed_runs']
+
+    @cached_property
+    def failed_course_runs(self):
+        """
+        Determine which course runs have been failed by the user.
+
+        Returns:
+            list of dicts, each a course run ID
+        """
+        return [run['course_run_id'] for run in self.course_runs_with_state['failed_runs']]
+
+    @cached_property
+    def course_runs_with_state(self):
+        """
+        Determine which course runs have been completed and failed by the user.
+
+        Returns:
+            dict with a list of completed and failed runs
+        """
         course_run_certificates = certificate_api.get_certificates_for_user(self.user.username)
-        return [
-            {'course_run_id': unicode(certificate['course_key']), 'type': certificate['type']}
-            for certificate in course_run_certificates
-            if certificate_api.is_passing_status(certificate['status'])
-        ]
+        completed_runs, failed_runs = [], []
+        for certificate in course_run_certificates:
+            course_data = {
+                'course_run_id': unicode(certificate['course_key']),
+                'type': certificate['type']
+            }
+            if certificate_api.is_passing_status(certificate['status']):
+                completed_runs.append(course_data)
+            else:
+                failed_runs.append(course_data)
+        return {'completed_runs': completed_runs, 'failed_runs': failed_runs}
```

Comment:
If you clean up the method name, this can just be `return {'completed': completed_runs, 'failed': failed_runs}`.